File: topKFrequent.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  let map = {};

  // Step 1: Count the frequency of each number
  for (let num of nums) {
    map[num] = (map[num] || 0) + 1;
  }

  // Step 2: Sort numbers based on frequency in descending order
  let sorted = Object.keys(map).sort((a, b) => map[b] - map[a]);

  // Step 3: Return the top k elements
  return sorted.slice(0, k).map(Number);
};

File: Python/topKFrequent.py
from collections import Counter
import heapq

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        # Step 1: Count frequency of each element
        freq_map = Counter(nums)  # like {'1': 3, '2': 2, '3': 1}

        # Step 2: Use heapq.nlargest to get the top k keys by frequency
        return heapq.nlargest(k, freq_map.keys(), key=freq_map.get)
